### ISSUE TITLE
Multi-version configuration files

### ISSUE THREAD

**Issue by CraigFe** [ocaml/dune#3101]
Currently, the config file in `$XDG_CONFIG_HOME/dune/config` sets a lower bound on the Dune version useable on one's machine (via the `lang` stanza):

```
File "/home/.config/dune/config", line 1, characters 11-14:
1 | (lang dune 2.1)
               ^^^
Error: Version 2.1 of dune is not supported.
Supported versions:
- 1.0 to 1.12
- 2.0
```

In order to access the latest features (such as the cache), I need to update _all_ projects on my machine in lockstep. I think it would be nice to allow projects to be upgraded independently while allowing access to the newest features. Here's two potential strategies:

----

### Proposal 1 - Multi-version config file

We could make all configuration settings forwards-compatible by adding some sort of conditional stanza:

```dune
(for-version 
 (>= 2.1.0)
 (cache enabled))
```

- *advantages*: retains a single source of truth for Dune config.
- *disadvantages*: makes it harder to change the config file parser in future, and makes the `lang` stanza in the config files mostly redundant.

----

### Proposal 2 - Multiple config files

We could have version-specific config files that are read in some sort of preference order, for instance:

```ocaml
$XDG_CONFIG_HOME/dune/config.2.0.1  (* Read by >= 2.0.1 *)
$XDG_CONFIG_HOME/dune/config.1.11.2 (* Read by >= 1.11.2 & < 2.01 *)
$XDG_CONFIG_HOME/dune/config        (* Read by < 1.11.2 *)
```

- *advantages*: logic for parsing the config is the same as before, and retains the same 'lang header' / 'body' parser distinction as the other Dune files.
- *disadvantages*: config parameters will likely be duplicated in the files.

----

Personally, I prefer the second option for not violating my expectations of the `lang` stanza. I'm happy to implement whatever solution is deemed best.

**Comment by Et7f3** [comment]
> config parameters will likely be duplicated in the files.
maybe you want to add a `(include config.1.11.2)`

**Comment by rgrinberg** [comment]
> maybe you want to add a (include config.1.11.2)

Indeed. Let's start with the simple thing here and just allow multiple config per version.
